### DISCUSSION TITLE
Feeling dense... just want to be able to on demand play a wav file stored in SPIFFS I2S to Internal DAC

### DISCUSSION THREAD

**Discussion by peteDDD** [pschatzmann/arduino-audio-tools#1734] [Q&A] [answered]
I have spent a few days playing with this and just am not getting it.  Feeling very dense (and tired).

I have an app where I occasionally need to play a wav file as an audible warning message.  (I do not want it continuously playing).   I want to use I2S and the internal DAC on an ESP-WROOM-32 (actually, a "Cheap Yellow Display" board).  It only has DAC_channel_2 (GPIO_26) going to the on-board single-channel analog amplifier.

Also, everything I have been able to get even a little "going" using I2S and the internal DAC leaves the DAC outputting a bit of noise all the time.   How can I "kill" the DAC between the times that I want to output my wav file so that there is no ambient noise coming out the speaker.  (If I don't start the I2S and DAC, I don't get any noise so I know the source of the noise is the active DAC).  (I have tried many things to do this but so far have not found any that work).

Is there an example of this that I can study? (I am using platform = espressif32@6.5.0)
MANY THANKS!

**Comment by pschatzmann** [comment, marked as answer]
Your description is quite confusing: either you use I2S with an extrernal DAC or the Analog API for the internal DAC!
If you are done with the output just call end to close the interface. Alternatively you can continue to send out silence...

I am also not sure with what you mean with espressif32@6.5.0: the latest IDF release is 5.3!

What sketch are you using ?
What development platform are you using ?

**Comment by peteDDD** [comment]
Well let's se if I can clarify... I think in Shatzman-speak, routing the
I2S to the internal DAC is what you CALL "Analog-API"

I have not found an example yet that does what I want to do but I imagine
it exists in your extensive catalog of examples.  I just have not found it
yet...

Objective: buffer using I2S, route to DAC, output through that path a wav
file on demand, not continuously.

I am using Platformio on Visual Studio Code and the most current rev of the
espressif32 platform is higher than 6.5 but that changed a lot of things
that broke a lot of my projects so I lock down to that rev for most of my
projects.

**Reply by pschatzmann** [reply]
The last releases with the __old__ I2S API (that also supported analog output) was in Arduino ESP 2.0.16 which used IDF 4.4.7
Since Arduino ESP 3.0 with IDF 5 there is a new I2S API for external DACs and separate analog APIs for the internal DAC and ADC!

In both cases you would just use the AnalogAudioStream in this library.

Please don't use PlatformIO version numbers: they don't mean anything to me. What is relevant is the related Arduino Version number or the related IDF version number !

I suggest to use Arduino Framework instead of espidf directly as described in the Wiki "How to use PlatformIO". You were not clear on this. 

In this case, there is no separate simple SPIFFS example necessary since this is just one of the many options how to manage files in Arduino with the ESP32. Just try to understand the chapter about Codecs.  

To make your sketch even simpler, I suggest to consider to use a MemoryStream and to convert and load your file to flash memory instead of using SPIFFS.

**Comment by peteDDD** [comment]
Hi Phil.  Thanks again for your reply.

Sorry if I caused confusion... regarding framework, I am using arduino
framework... in platformio.ini:
  platform = ***@***.***
  board = esp32dev
  framework = arduino

I will definitely study the Codecs section.  Thank you for that pointer.

I understand the benefits of using a MemoryStream for the sound data for
many applications.  In the case of the current application that I am
working on, we have a concrete reason for keeping the wav files in SPIFFS
and separate from the code.   In managing our over-the-air updates, we can
choose to push either an update to files saved in SPIFFS or an update to
the firmware. Separating the language-specific files (including the sound
files) allows us to reduce over-the-air update times significantly.  Also,
maintaining the language-specific files (including sound files) in the
SPIFFS space allows us to manage country or language specific units in the
field based on the SPIFFS files while maintaining a single version of the
core firmware across the entire fleet devices.

Thanks again,
Pete

**Comment by peteDDD** [comment]
I have a solution, though it seems more complex than it should...

first, I solved the issue of a slight hum from the internal DAC output.  After trying all of these things...
```
  player.end();  //does not change the hum

  player.setSilenceOnInactive();  //only makes it worse (more of a buzz, than a slight hum)

  player.stop(); //makes it even worse, a click every few seconds.

  player.setVolume(0.0); //does not change the hum
```

I finally settled on disabling the DAC after the wav file completes playing by using:

`dac_output_disable(DAC_CHANNEL_2);  // WORKS.  Silences DAC output; no hum.   with #include <driver/dac.h>`

and re-enabling the DAC each time I play a wav file (see playWAV below)

(the Cheap Yellow Board only has one audio output channel... DAC_CHANNEL_2)

This led me to creating two functions:
`  void playWAV(const char *fileFilterPattern)`
and
`  bool checkIfWavIsDonePlaying()`
and a bool flag
`  bool b_AskedToPlayWav`
since we only should check player.isActive if we are in a state where a wav has actually been started since player.isActive will return false if a wav has completed or if a wav has not been started.

In this sample code, I just play the next file in a list of files.  In "real life" there are other things I would do after the wav playing was completed.  The point is that I wanted a way to play selected wav files from SPIFFS on demand with the loop() of my code and have the speaker absolutely quiet the rest of the time.

I am VERY WELCOME to getting input for simpler ways of doing this remembering the objectives:
1.  need to use wav files from SPIFFS
2.  need to be able to play selected wav files on demand from within the loop()
3.  need to completely silence the DAC whenever the wavs are not being played.

So, here is the test code...
```
#include "AudioTools.h"
#include "AudioTools/AudioLibs/AudioSourceSPIFFS.h"
#include "AudioTools/AudioCodecs/CodecWAV.h"

#include <driver/dac.h>

const char *startFilePath = "/";
const char *ext = "wav";

// list of wav files to play, saved in data folder
const char *wavFiles[] = {"*earth_move*", "*tremble*"};
const int numOfWavFiles = sizeof(wavFiles) / sizeof(wavFiles[0]);
uint8_t currentWavFileIndex = 0;
bool b_AskedToPlayWav = false;

AudioSourceSPIFFS source(startFilePath, ext);
AnalogAudioStream out;
WAVDecoder decoder;
AudioPlayer player(source, out, decoder);


void playWAV(const char *fileFilterPattern)
{
  dac_output_enable(DAC_CHANNEL_2);
  source.setFileFilter(fileFilterPattern);
  b_AskedToPlayWav = true;
  player.begin();
}

bool checkIfWavIsDonePlaying()
{
  if (!player.isActive())
  {
    b_AskedToPlayWav = false;          // done playing wave
    dac_output_disable(DAC_CHANNEL_2); // silence the DAC output
    return true;  // done playing wav
  }
  else
  {
    player.copy();
    return false; // still playing wav
  }
}

void setup()
{
  // setup output
  auto cfg = out.defaultConfig();
  out.begin(cfg);

  player.setVolume(0.5);
 
  playWAV(wavFiles[currentWavFileIndex]); // start playing first wav file
}

void loop()
{
  if (b_AskedToPlayWav)
  {
    if (checkIfWavIsDonePlaying())
    {
      // what do we want to do after the wav is done playing?
      //   let's play the next wave file in the array
      //   until there are no more files to play
      if (currentWavFileIndex++ < numOfWavFiles - 1)
      {
        playWAV(wavFiles[currentWavFileIndex]); // play the next WAV file
      }
    }
  }

  // do other stuff here...
}
```

**Reply by pschatzmann** [reply]
If you want to control the sequence you might be better off using a AudioSourceCallback class were you just return the Files in the required sequence (see this [example](https://github.com/pschatzmann/arduino-audio-tools/blob/aeda11b96e23a09f9013e3fe92e1c4c8b97d201a/examples/examples-player/player-callback-i2s/player-callback-i2s.ino#L25)). There you can also  manage the special logic to close the output when there are no more files.

You can easily end the AnalogAudioStream by calling end() and when you want to restart just call begin() again.

ps. There is a big chance that your noise on the DAC is comping from a noisy power supply. 
If you are just playing individual files you might be better off by just using a bare bone EncodedAudioStream instead of the player.

**Comment by peteDDD** [comment]
If I understand EncodedAudioStream, as in this example: stream-memory_wav-pwm.ino, then I have to provide the size and bit depth of each file into the code whereas the player using WAVdecoder figures that out for itself.  That of course helps the goal of keeping the field deployment/updating easy.

Towards you advice to "end" the AnalogAudioStream, if I understand correctly, where I have used AnalogAudioStream out; then I would just use out.end(); correct?  I avoid opening and closing (beginning/ending) objects repeatedly for fear of unknown memory leaks catching up with me, but I'll give it a go.

In re noisy power supply, that certainly is likely but not something that can be fixed to the level of total silence in the application.  The nature of the power in the countries were we will deploy these devices is deplorable to start with and to the extent that the noisy power does not otherwise effect the processor and display, disabling the DAC while not playing the waves is a very workable solution.

the AudioSourceCallback is nice but I playing the files in sequence was just used in my test code to show that I could call different files, not necessarily in a set sequence.  In fact, in the app, the objective is to play sounds (announcements) based on the state of the system...

You have created a very rich tool here which I imagine I will revisit many times over the years for various projects.  I am just nibbling off a bit for a specific project now.  Thanks again!

**Reply by pschatzmann** [reply]
Nop, you can just call copy from the file to the EncodedAudioStream like in [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-stream/streams-sd_mp3-i2s/streams-sd_mp3-i2s.ino)

**Comment by peteDDD** [comment]
Thanks Phil.  That's great and it runs a bit faster.

I appreciate your patience as I learned this.   I am just stuck on one last thing... how to set the volume. And I just found this example https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control so I will work through that...

Here is the code I have now.
```
#include <Arduino.h>

#include "AudioTools.h"
#include "AudioTools/AudioLibs/AudioSourceSPIFFS.h"
#include "AudioTools/AudioCodecs/CodecWAV.h"

#include "driver/dac.h"

// list of wav files to play, saved in data folder
const char *wavFiles[] = {"/earth_move.wav", "/tremble.wav", "/earth_move.wav"};
const int numOfWavFiles = sizeof(wavFiles) / sizeof(wavFiles[0]);
uint8_t currentWavFileIndex = 0;
bool b_AskedToPlayWav = false;

AnalogAudioStream out;
EncodedAudioStream decoder(&out, new WAVDecoder()); // Decoding stream
StreamCopy copier;
File audioFile;


void playWAV(const char *fileName)
{
  audioFile = SPIFFS.open(fileName, "r");
  b_AskedToPlayWav = true;
  dac_output_enable(DAC_CHANNEL_2);
  copier.begin(decoder, audioFile);
}

bool checkIfWavIsDonePlaying()
{
  if (!copier.copy())
  {
    b_AskedToPlayWav = false; // done playing wave
    dac_output_disable(DAC_CHANNEL_2); // silence the DAC output
    return true;              // done playing wav
  }
  else
  {
    return false; // still playing wav
  }
}

void setup()
{
  Serial.begin(115200);
  SPIFFS.begin();

  // setup output
  auto outCfg = out.defaultConfig();
  out.begin(outCfg);

  //! add setting of volume here

  decoder.begin();

  playWAV(wavFiles[currentWavFileIndex]);
}

void loop()
{
  if (b_AskedToPlayWav)
  {
    if (checkIfWavIsDonePlaying())
    {
      if (currentWavFileIndex++ < numOfWavFiles - 1)
      {
        Serial.println("PLAY NEXT FILE");  // shows when we actually start the playWAV call
        playWAV(wavFiles[currentWavFileIndex]); // play the next WAV file
      }
    }
  }

  // do other stuff here...
}
```

**Reply by peteDDD** [reply]
As we say, here in the highlands of Panama, **_Listo!_**
I think I finally got the hang of this very powerful library.
Thanks again!

```
#include <Arduino.h>

#include "AudioTools.h"
#include "AudioTools/AudioLibs/AudioSourceSPIFFS.h"
#include "AudioTools/AudioCodecs/CodecWAV.h"

#include "driver/dac.h"

// list of wav files to play, saved in data folder
const char *wavFiles[] = {"/earth_move.wav", "/tremble.wav", "/earth_move.wav"};
const int numOfWavFiles = sizeof(wavFiles) / sizeof(wavFiles[0]);
uint8_t currentWavFileIndex = 0;
bool b_AskedToPlayWav = false;


// StreamCopy audioFile -> EncodedAudioStream decoder -> VolumeStream volume -> DACout
AnalogAudioStream DACout;
VolumeStream volume(DACout);
EncodedAudioStream decoder(&volume, new WAVDecoder()); // Decoding stream
StreamCopy copier;
File audioFile;


void playWAV(const char *fileName)
{
  audioFile = SPIFFS.open(fileName, "r");
  b_AskedToPlayWav = true;
  dac_output_enable(DAC_CHANNEL_2);
  copier.begin(decoder, audioFile);
}

bool checkIfWavIsDonePlaying()
{
  if (!copier.copy())
  {
    b_AskedToPlayWav = false; // done playing wave
    dac_output_disable(DAC_CHANNEL_2); // silence the DAC output
    return true;              // done playing wav
  }
  else
  {
    return false; // still playing wav
  }
}

void setup()
{
  Serial.begin(115200);
  SPIFFS.begin();

  // setup output
  auto DACoutCfg = DACout.defaultConfig();
  DACout.begin(DACoutCfg);

  // set initial volume  (ref: https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control)
  volume.begin(DACoutCfg); // we need to provide the bits_per_sample and channels
  volume.setVolume(0.5); // accepts floatvalues0.0 to 1.0

  decoder.begin();

  playWAV(wavFiles[currentWavFileIndex]);
}

void loop()
{
  if (b_AskedToPlayWav)
  {
    if (checkIfWavIsDonePlaying())
    {
      if (currentWavFileIndex++ < numOfWavFiles - 1)
      {
        Serial.println("PLAY NEXT FILE");  // shows when we actually start the playWAV call
        playWAV(wavFiles[currentWavFileIndex]); // play the next WAV file
      }
    }
  }

  // do other stuff here...
}
```

**Reply by pschatzmann** [reply]
Right: that's the best way to handle the volume

**Reply by peteDDD** [reply]
"There is a chapter in the Wiki that explains all options for the Volume control"

There is SO MUCH documentation, in so many places...
